### PR TITLE
relevant themes(#816) and delete improvement

### PR DESCRIPTION
**Pull request by optizone8**

### CODE REVIEW

File: src/common/Configuration.h [hunk 1]
```diff
@@ -8,6 +8,10 @@
 #define Config() (Configuration::instance())
 #define ConfigColor(x) Config()->getColor(x)
 
+constexpr int CONFIG_DEFAULT_THEME = 0;
+constexpr int CONFIG_DARK_THEME = 1;
+//constexpr int CONFIG_LIGHT_THEME = 2;
```
Comment (xarkes) [inline src/common/Configuration.h]: Better use an enum.

File: src/dialogs/preferences/AppearanceOptionsWidget.cpp [hunk 2]
```diff
@@ -14,6 +14,39 @@
 #include "common/ColorSchemeFileSaver.h"
 #include "widgets/ColorSchemePrefWidget.h"
 
+
+//static constexpr int lightTheme = 4; // If we finally add light theme
+static constexpr int defaultTheme = 1;
+static constexpr int darkTheme = 2;
```
Comment (xarkes) [inline src/dialogs/preferences/AppearanceOptionsWidget.cpp]: Why are you redefining this? It is already in Configuration.h

File: src/dialogs/preferences/AppearanceOptionsWidget.cpp [hunk 3]
```diff
@@ -14,6 +14,39 @@
...
+
+
+bool shouldShow(const QString &s)
+{
+    static const QMap<QString, int> relevantSchemes = {
+        //       { "default", darkTheme },
```
Comment (xarkes) [inline src/dialogs/preferences/AppearanceOptionsWidget.cpp]: Remove dead code.

File: src/dialogs/preferences/AppearanceOptionsWidget.cpp [hunk 4]
```diff
@@ -49,11 +85,16 @@ void AppearanceOptionsWidget::updateThemeFromConfig()
 
     QList<QString> themes = Core()->getColorThemes();
     ui->colorComboBox->clear();
-    ui->colorComboBox->addItem("default");
+    //    ui->colorComboBox->addItem("default");
```
Comment (xarkes) [inline src/dialogs/preferences/AppearanceOptionsWidget.cpp]: Deadcode.

File: src/dialogs/preferences/AppearanceOptionsWidget.cpp [hunk 5]
```diff
@@ -49,11 +85,16 @@ void AppearanceOptionsWidget::updateThemeFromConfig()
...
     for (QString str : themes)
```
Comment (xarkes) [inline src/dialogs/preferences/AppearanceOptionsWidget.cpp]: Better add `{` here on the for and if for better readability.

File: src/dialogs/preferences/AppearanceOptionsWidget.cpp (hunk 2 above)
Comment (optizone8) [reply src/dialogs/preferences/AppearanceOptionsWidget.cpp]: It's flags. For instance, if we decide that some theme relevant for both Qt themes, we just add line `{ "name", defaultTheme | darkTheme },` and one will appear in both situations. Same when we add extra Qt theme. Just for flexibility.

Comment (xarkes) [reply src/dialogs/preferences/AppearanceOptionsWidget.cpp]: Yep but I'm pretty sure you can mix that up with the stuff you added in the .h already, no?
If not move that in the header anyway and use a proper name.

File: src/dialogs/preferences/AppearanceOptionsWidget.cpp [hunk 6]
```diff
@@ -105,12 +149,21 @@ void AppearanceOptionsWidget::on_copyButton_clicked()
...
 void AppearanceOptionsWidget::on_deleteButton_clicked()
 {
-    ColorSchemeFileWorker().deleteScheme(Config()->getCurrentTheme());
+    if (ColorSchemeFileWorker().isCustomScheme(Config()->getCurrentTheme())) {
+        QMessageBox mb;
+        mb.setWindowTitle(tr("Delete"));
+        mb.setText(tr("Are you sure you want to delete theme ") + Config()->getCurrentTheme());
+        mb.setIcon(QMessageBox::Question);
+        mb.setStandardButtons(QMessageBox::Yes | QMessageBox::No);
+        if (mb.exec() == QMessageBox::Yes) {
```
Comment (a1ext) [inline src/dialogs/preferences/AppearanceOptionsWidget.cpp]: Why are you using so much code (object creation `mb`, many assignments) to do a simple thing?
It have to be as short and readable as possible, something like the following:
```cpp
if (QMessageBox::Yes == QMessageBox::question(nullptr,
                                              tr("Delete"),
                                              tr("Are you sure you want to delete theme ") + Config()->getCurrentTheme(),
                                              QMessageBox::Yes | QMessageBox::No)) {
    // do smth here...
}
```

File: src/dialogs/preferences/AppearanceOptionsWidget.cpp (hunk 2 above)
Comment (optizone8) [reply src/dialogs/preferences/AppearanceOptionsWidget.cpp]: Yeah, You're right. I forgot that I can assign values in enum

File: src/dialogs/preferences/AppearanceOptionsWidget.cpp [hunk 7]
```diff
@@ -24,12 +57,15 @@ AppearanceOptionsWidget::AppearanceOptionsWidget(PreferencesDialog *dialog, QWid
...
+    connect(Config(), &Configuration::fontsUpdated, this,
+            &AppearanceOptionsWidget::updateFontFromConfig);
+    connect(ui->colorComboBox,
+            static_cast<void (QComboBox::*)(const QString &)>(&QComboBox::currentIndexChanged),
+    [&](const QString & name) {
+        static_cast<ColorSchemePrefWidget *>(ui->colorSchemePrefWidget)->setNewScheme(name);
     });
     static_cast<ColorSchemePrefWidget *>
-    (ui.get()->colorSchemePrefWidget)->setNewScheme(Config()->getCurrentTheme());
+    (ui->colorSchemePrefWidget)->setNewScheme(Config()->getCurrentTheme());
```
Comment (a1ext) [inline src/dialogs/preferences/AppearanceOptionsWidget.cpp]: Is this (`static_cast<ColorSchemePrefWidget *>`) cast mandatory here and in all other places?

File: src/dialogs/preferences/AppearanceOptionsWidget.cpp (hunk 6 above)
Comment (optizone8) [reply src/dialogs/preferences/AppearanceOptionsWidget.cpp]: Didn't know. Will change.

File: src/dialogs/preferences/AppearanceOptionsWidget.cpp (hunk 7 above)
Comment (optizone8) [reply src/dialogs/preferences/AppearanceOptionsWidget.cpp]: Yes, otherwise I can't call function members of this class

Comment (a1ext) [reply src/dialogs/preferences/AppearanceOptionsWidget.cpp]: Could you explain? the type of it doesn't change after the cast, why are you cannot call it?

Comment (optizone8) [reply src/dialogs/preferences/AppearanceOptionsWidget.cpp]: Oh, apparently at some point of time it was necessary, but now it outdated and I just kept it without the reason.

File: src/dialogs/preferences/AppearanceOptionsWidget.cpp [hunk 8]
```diff
@@ -24,12 +51,15 @@ AppearanceOptionsWidget::AppearanceOptionsWidget(PreferencesDialog *dialog, QWid
     updateFontFromConfig();
     updateThemeFromConfig();
 
-    connect(Config(), &Configuration::fontsUpdated, this, &AppearanceOptionsWidget::updateFontFromConfig);
-    connect(ui.get()->colorComboBox, &QComboBox::currentTextChanged, [&](const QString & name) {
-        static_cast<ColorSchemePrefWidget *>(ui.get()->colorSchemePrefWidget)->setNewScheme(name);
+    connect(Config(), &Configuration::fontsUpdated, this,
+            &AppearanceOptionsWidget::updateFontFromConfig);
+
+    connect(ui->colorComboBox, &QComboBox::currentTextChanged,
```
Comment (a1ext) [inline src/dialogs/preferences/AppearanceOptionsWidget.cpp]: Redundant lambda usage. Have to be simplified to:
```cpp
connect(ui->colorComboBox, &QComboBox::currentTextChanged,
    ui->colorSchemePrefWidget, &ColorSchemePrefWidget::setNewScheme);
```

File: src/common/Configuration.h [hunk 9]
```diff
@@ -8,6 +8,12 @@
 #define Config() (Configuration::instance())
 #define ConfigColor(x) Config()->getColor(x)
 
+enum CutterQtThemes {
+    defaultTheme = 1,
+    darkTheme = 2,
+    lightTheme = 4
```
Comment (ITAYC0HEN) [inline src/common/Configuration.h]: I don't think we have "Light" theme (yet). This also mentioned in other places on your PR.
Is this for future use?

Comment (optizone8) [reply src/common/Configuration.h]: yeah, for future.

**Comment by ITAYC0HEN** [comment]
I fetched your PR and tested it. I encountered several bugs, including a segfault (which might be unrelated but it happened when I tried your feature)

First, upon opening Cutter with your branch, this was how it looks:

image


This is obvioulsy wrong, since you can see that there is a mix of the "ayu" color theme in the disassembly as well as the bright QT theme. Moreover, I always use the Dark theme (usually with `ayu` color theme) and this means that your PR forcibly changed my QT Theme to "Default".

Upon checking the "Preferences" settings, I saw that there is a Color Theme i never saw before, called "ayu2" -- and that it is available on the "Default" QT Theme. Shouldn't be there.

image


Next thing I noticed, is that whenver I change my QT Theme, from Default to Dark back and forth, the Color theme returns to default and not to the one was before.




I have also reset my Settings and the problems occur - more than that - please look careful at the following GIF:



As you can see, upon open the "settings" window the theme is going mad and there's only one available theme which is the ayu.

---

**Bottom line --**

Your PR is broken and, sadly, can't be merged at this time. Please address the mentioned bugs, find others and test it thoroughly. Please develop this feature from the eyes of the use and imagine how would you expect it to be.

**Comment by ITAYC0HEN** [comment]
Thank you for the improvmetns :)
As you mentioned, you did not succeed in preventing the dark color-theme to be shown with the Light QT theme upon the first execution of Cutter. Sadly, this behavior is unacceptable and we would need to find an elegant solution to this before merging this PR. There is no reason to modify the user settings and rollback to Default  QT theme if my settings are set to Dark.

image
_Image: dark color theme (ayu), with light QT theme_

As you can see in the image below, the PR again created the "ayu2" theme for me and made it available under the Default QT Theme. Again, this should be fixed.
image

**Comment by optizone8** [comment]
If it is available on both QT Themes, it is maybe your own theme, because as I mentioned only hardcoded and user's themes are available to display.

**Comment by ITAYC0HEN** [comment]
Steps to reproduce:
1. Build Cutter from master
2. Choose QT dark theme and Ayu color schene
3. Checkout your branch
4. Look at the color themes

These steps creates "ayu2" for me. I have never created it myself

**Comment by optizone8** [comment]
About "ayu2". I couldn't reproduce this bug neither on last commit nor  "add themes" one. 
May you show me output of `Core()->getColorThemes();` and `ColorSchemeFileWorker().isCustomScheme("ayu2")` inside of member `AppearanceOptionsWidget::updateThemeFromConfig()`, it'll very help me in my investigating.

File: src/common/Configuration.cpp [hunk 10]
```diff
@@ -48,6 +48,9 @@ static const QHash<QString, QVariant> asmOptions = {
 Configuration::Configuration() : QObject()
 {
     mPtr = this;
+    if (s.contains("ColorPalette") && !s.contains("Not first start after #846")) {
+        setTheme(getTheme() + 1);
```
Comment (a1ext) [inline src/common/Configuration.cpp]: Why do you use numerical theme ID?

**Comment by optizone8** [comment]
Ok, I'll explain the issue.
First, why did I change **defaultTheme** and **darkTheme** ID's? To have possibility write this type of line:
`{ "some theme", defaultTheme | darkTheme }` so this theme was displayed on both Qt Themes.
Same with Qt Themes we haven't add yet: 
`{ "some theme", defaultTheme | darkTheme | someQtThemeWeHaven'tAddYet }`.

Of course, I can change back **defaultTheme** and **darkTheme** ID's to 0 ad 1 instead of 1 and 2. But this action will make it impossible to configure displaying color themes so flexible. 

What the problem is?
When first start after this PR, dark theme changes to default, default - to dark.
Ways to solve it:
1) Change back **defaultTheme** and **darkTheme** ID's
2) Delete/change settings file during building
3) Use this solution
4) Use solution I'll commit soon

**Comment by ITAYC0HEN** [comment]
ping @optizone8 :)
any update on the progress?

**Comment by optizone8** [comment]
I proposed seweral ways to solve issue. And now I wait your thoughts which one I should use.

**Comment by ITAYC0HEN** [comment]
Works great now, thank you!!
